{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1ebb41f8",
   "metadata": {},
   "source": [
    "# Reproducing a Paper - Softconsensus\n",
    "\n",
    "In this section, we will reproduce the functionality described in  [[von LÃ¶hneysen et al. - 2023]](https://journals.sagepub.com/doi/pdf/10.1089/cmb.2024.0519). Below is a brief introduction to the key concepts required for this task. For further details, please refer to the original paper.\n",
    "\n",
    "## Introduction \n",
    "As discussed in the first part, covariation information from a multiple sequence alignment can improve the folding of the alignment. This is done by **RNAalifold**, which outputs a consensus structure for the alignment. However, the consensus structure cannot always be directly applied to individual sequences. To address this, we will implement a pipeline to construct a secondary structure for an individual sequence of the alignment, which we will name focal sequence from now on.\n",
    "\n",
    "### Strategy I: Hard Constraints from the Consensus Structure\n",
    "\n",
    "In this strategy, we will extract all valid base pairs from the consensus structure generated by RNAalifold and use them as hard constraints to refold the focal sequence. In order to filter out nonsensical interactions, the consensus structure must first be modified by:\n",
    "\n",
    "1. Removing all base pairs with non-compatible nucleotides in the focal sequence.  \n",
    "2. Removing all base pairs where one interaction partner maps to a gap position in the focal sequence.  \n",
    "3. Removing all hairpins with a length less than 3.  \n",
    "\n",
    "This approach ensures that the constraints are valid in the context of the focal sequence while adopting as many interactions from the consensus structure as possible.\n",
    "\n",
    "### Strategy II: Base Pair Probability and Soft Constraints\n",
    "\n",
    "Here, we will leverage base pair probabilities and soft constraints to refine the folding process. This is achieved by assigning a bonus energy to each valid base pair in the consensus structure. This bonus energy of an interaction is selected to be proportional to its base pairing probability within the ensemble.\n",
    "\n",
    "More specifically, the bonus energy for a feature $\\gamma$ (such as a base pair) is calculated as the difference between the free energies of the ensemble with and without the feature $\\gamma$. This approach allows us to incorporate probabilistic information from the ensemble into the folding process, providing a more nuanced and flexible folding improvement compared to hard constraints."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ffaacb6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Installing all requirements using conda. This method is recommended if you want to follow this tutorial on your own machine. If you have already created a conda environment for the first Notebook, you can use that.\n",
    "!conda create --yes --name ViennaRNA-Tutorial python=3.14\n",
    "!conda activate ViennaRNA-Tutorial\n",
    "!conda install --yes bioconda::viennarna\n",
    "!conda install --yes matplotlib\n",
    "!conda install --yes numpy\n",
    "!conda install --yes ipykernel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1901c577",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Run this cell to set up the requirements on Google Colab. Don't run it on your local machine!\n",
    "#Furthermore, you need to create a folder called 2_input in the Colab Working Directory (at /content/2_input) and upload the given files.\n",
    "!pip install ViennaRNA\n",
    "!apt-get update -qq\n",
    "!apt-get install -y vienna-rna\n",
    "!pip install cairosvg"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6a065fa8",
   "metadata": {},
   "source": [
    "Here are a few helper functions that will be used throughout the tutorial. Simply run the next four cells to load them into your environment."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "765e400a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import re\n",
    "\n",
    "import RNA\n",
    "import numpy as np\n",
    "from IPython.display import Image, HTML, display\n",
    "import cairosvg\n",
    "\n",
    "\n",
    "# The path to the provided alignment file\n",
    "alignment_path = \"2_input/alignment.fa\"\n",
    "\n",
    "# The reference structure of the focal sequence we want to reproduce\n",
    "focal_seq_reference_ss = \"((((((((((.....((((((((....(((((((.............))))..)))...)))))).)).((.((....((((((((...))))))))....)).))...)))))))))).\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c58ab68d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_structures(ss):    \n",
    "    \"\"\"This functions plots an RNA structure and displays it next to the focal_seq_reference_ss.\n",
    "\n",
    "    Args:\n",
    "        ss (str): The secondary structure to display as a dot bracket string.\n",
    "    \"\"\"\n",
    "    \n",
    "    i = 1\n",
    "    while os.path.exists(f\"2_input/result_plot{i}.svg\"):\n",
    "        i += 1\n",
    "\n",
    "    reference_path = \"2_input/reference_plot.svg\"\n",
    "    result_path = f\"2_input/result_plot{i}.svg\"\n",
    "\n",
    "    RNA.svg_rna_plot(' '*len(focal_seq_reference_ss), focal_seq_reference_ss, reference_path)\n",
    "    RNA.svg_rna_plot(' '*len(ss), ss, result_path)\n",
    "\n",
    "    cairosvg.svg2png(url=reference_path, write_to=reference_path.replace(\".svg\", \".png\"))\n",
    "    cairosvg.svg2png(url=result_path, write_to=result_path.replace(\".svg\", \".png\"))\n",
    "\n",
    "    print(\"Reference:\")\n",
    "    display(Image(filename=reference_path.replace(\".svg\", \".png\")))\n",
    "    print(\"Result:\")\n",
    "    display(Image(filename=result_path.replace(\".svg\", \".png\")))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "a10c099c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_alignment(path):\n",
    "    \"\"\"Reads the alignment file in fasta format and returns a list containing all sequences.\n",
    "\n",
    "    Args:\n",
    "        path (str): Path to the alignment in fasta format\n",
    "\n",
    "    Returns:\n",
    "        list: A list containing all sequences of the alignment\n",
    "    \"\"\"\n",
    "\n",
    "    aligned_seq = []\n",
    "    with open(path, 'r') as inF:\n",
    "        for line in inF:\n",
    "            if line.startswith('>'):\n",
    "                continue\n",
    "            aligned_seq.append(line.strip().upper().replace('T', 'U'))\n",
    "    return aligned_seq"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11ac1a09",
   "metadata": {},
   "source": [
    "The next cell defines all the input variables required for the task. These include:\n",
    "\n",
    "- **aligned_seq**: A list of all aligned sequences, including gap characters.  \n",
    "- **seq**: A list of all aligned sequences without gap characters. The lengths of these sequences vary due to the removal of gaps.  \n",
    "- **focal_seq**: The focal sequence (In this case simply the first sequence in the alignment). The primary goal is to predict the secondary structure of this sequence using information from the alignment. This sequence does not contain gap characters.  \n",
    "- **aligned_focal_seq**: The focal sequence as it appears in the alignment, including gap characters.  \n",
    "- **compatible_bases**: A list of all combinations of bases that can form canonical base pairs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "7e655263",
   "metadata": {},
   "outputs": [],
   "source": [
    "aligned_seq = read_alignment(alignment_path)\n",
    "seq = [s.replace('-', '') for s in aligned_seq]\n",
    "focal_seq = seq[0]\n",
    "aligned_focal_seq = aligned_seq[0]\n",
    "compatible_bases = [\"AU\", \"UA\", \"CG\", \"GC\", \"UG\", \"GU\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "804b53c5",
   "metadata": {},
   "source": [
    "## MFE folding\n",
    "\n",
    "Now it's your turn: Perform a basic sanity check by folding the focal sequence using the standard MFE algorithm without integrating the additional information from the alignment.  Use the function `plot_structures(ss)` to compare the resulting structure to the ground truth by visualizing them."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c1ed1798",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "f83fc884",
   "metadata": {
    "type": "solution"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "UGCCUGGCGGCCGUAGCGCGGUGGUCCCACCUGACCCCAUGCCGAACUCAGAAGUGAAACGCCGUAGCGCCGAUGGUAGUGUGGGGUCUCCCCAUGCGAGAGUAGGGAACUGCCAGGCAU\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr><td><img src='2_input/reference_plot.svg'></td><td><img src='2_input/result_plot7.svg'></td></tr>\n",
       "            <tr><td style='text-align: center;'>Reference</td><td style='text-align: center;'>Result</td></tr>\n",
       "        </table>\n",
       "    "
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fc = RNA.fold_compound(focal_seq)\n",
    "ss, mfe = fc.mfe()\n",
    "print(focal_seq)\n",
    "plot_structures(ss)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d7411058",
   "metadata": {},
   "source": [
    "## Strategy I implementation (hard constraint)\n",
    "Next, we aim to improve the result using *Strategy I*, as described above. Start by computing the consensus structure of the alignment stored in `aligned_seq`. Create a fold compound using the alignment and then calculate the MFE consensus structure."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13a04ba8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "71da926b",
   "metadata": {
    "type": "solution"
   },
   "outputs": [],
   "source": [
    "fc_aln = RNA.fold_compound(aligned_seq)\n",
    "\n",
    "consensus_ss, consensus_mfe = fc_aln.mfe()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cdbc6b67",
   "metadata": {},
   "source": [
    "The resulting secondary structure incorporates information from all sequences in the alignment, making it a more reliable prediction. However, it may not be fully compatible with the single focal sequence we aim to fold.\n",
    "\n",
    "To address this, we will remove all interactions in the consensus structure that are incompatible with the focal sequence. Specifically, we will delete interactions that:\n",
    "\n",
    "- Form non-canonical base pairs\n",
    "- Include a gap character in the base pair\n",
    "- Create hairpins with a length < 3\n",
    "\n",
    "Save the cleaned secondary structure in dot bracket notation.\n",
    "\n",
    "*Hint: Use `aligned_focal_seq` and the consensus structure you computed earlier, which should have the same length. Representing the structure as a pair table can simplify the removal of non-canonical pairs and interactions with gap characters. However, keep in mind that a pair table is 1-indexed (the 0th element stores the structure length), while the sequence is 0-indexed.*\n",
    "\n",
    "*For the fixed hairpin length cutoff of 3, simple string replacement will suffice.*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ebe8a1c3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "1683e5cb",
   "metadata": {
    "type": "solution"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "UGCCUGGCGG-CCGUAGCGCG-G-UGGUCCC-AC-CUGACCCCAUGCCGAACUCA-GA-AGUGAAACGCCGUA-GC-G--CCGAUG----GUAGUGUG--GGGUCUCCCCAUGCGAG--AGUA-GG-G-AA-CUGCCAGGCAU\n",
      "..((((((((.(....(((((.(.((.........((((.............))).).........))))))..)).(..((.((........(((((..((....)).)))))......))..)).)..).))))))))...\n",
      "\n",
      "\n",
      "CA\n",
      "GG\n",
      "UGCCUGGCGGCCGUAGCGCGGUGGUCCCACCUGACCCCAUGCCGAACUCAGAAGUGAAACGCCGUAGCGCCGAUGGUAGUGUGGGGUCUCCCCAUGCGAGAGUAGGGAACUGCCAGGCAU\n",
      "..((((((((.....((((((((.......((((.............))))........)))))).)).((.((....(((((((....)).)))))....)).))...))))))))...\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "consensus_ss_pt = RNA.ptable(consensus_ss)\n",
    "clean_ss = \"\"\n",
    "print(f\"{aligned_focal_seq}\\n{consensus_ss}\\n\\n\")\n",
    "\n",
    "#Step1: remove all gap positions and non-canonical interactions\n",
    "for nuc, ss_idx in zip(aligned_focal_seq, range(1, consensus_ss_pt[0]+1)):\n",
    "    if nuc == '-':\n",
    "        if consensus_ss_pt[ss_idx] > 0:\n",
    "            consensus_ss_pt[consensus_ss_pt[ss_idx]] = 0\n",
    "        continue\n",
    "    if consensus_ss_pt[ss_idx] == 0:\n",
    "        clean_ss += '.'\n",
    "    elif consensus_ss_pt[ss_idx] > ss_idx:\n",
    "        if f\"{nuc}{aligned_focal_seq[consensus_ss_pt[ss_idx]-1]}\" not in compatible_bases:\n",
    "            print(f\"{nuc}{aligned_focal_seq[consensus_ss_pt[ss_idx]-1]}\")\n",
    "            consensus_ss_pt[consensus_ss_pt[ss_idx]] = 0\n",
    "            clean_ss += '.'\n",
    "        else:\n",
    "            clean_ss += '('\n",
    "    elif consensus_ss_pt[ss_idx] < ss_idx:\n",
    "        clean_ss += ')'\n",
    "\n",
    "#Step2: remove hairpins of length < 3\n",
    "clean_ss.replace(\"()\", \"..\").replace(\"(.)\", \"...\").replace(\"(..)\", \"....\")\n",
    "\n",
    "print(f\"{focal_seq}\\n{clean_ss}\\n\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c3a335d",
   "metadata": {},
   "source": [
    "When you have cleaned the secondary structure, use it as a hard constraint to fold the focal sequence. Visualize the results using the `plot_structures(ss)` function provided earlier.\n",
    "\n",
    "*Hint: Use the `fc.constraints_add()` function to apply the constraints. Experiment with different flags to weakly or strongly enforce the constraints and observe if the results change.*\n",
    "\n",
    "*You can also weaken the constraints by not enforcing the exact base pairs but instead just forcing the bases to form any interaction. In the pseudo dot bracket notation used in the ViennaRNA Package, this type of constraint is represented by the character `|`*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49653de4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "eb2ed1a6",
   "metadata": {
    "type": "solution"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "..((((((((.....((((((((.......((((.............))))........)))))).)).((.((....(((((((....)).)))))....)).))...))))))))... (constraint)\n",
      "((((((((((.....((((((((....(((((((.............))))..)))...)))))).)).((.((....(((((((....)).)))))....)).))...)))))))))). (MFE)\n",
      "((((((((((.....((((((((....(((((((.............))))..)))...)))))).)).((.((....((((((((...))))))))....)).))...)))))))))). (target)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr><td><img src='2_input/reference_plot.svg'></td><td><img src='2_input/result_plot4.svg'></td></tr>\n",
       "            <tr><td style='text-align: center;'>Reference</td><td style='text-align: center;'>Result</td></tr>\n",
       "        </table>\n",
       "    "
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fc = RNA.fold_compound(focal_seq)\n",
    "fc.constraints_add(clean_ss, RNA.CONSTRAINT_DB_DEFAULT)\n",
    "#fc.constraints_add(clean_ss, RNA.CONSTRAINT_DB_DEFAULT|RNA.CONSTRAINT_DB_ENFORCE_BP)\n",
    "\n",
    "ss, mfe = fc.mfe()\n",
    "print(clean_ss, \"(constraint)\")\n",
    "print(ss, \"(MFE)\")\n",
    "print(focal_seq_reference_ss, \"(target)\")\n",
    "\n",
    "plot_structures(ss)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c677af97",
   "metadata": {},
   "source": [
    "The first strategy resulted in a significant improvement compared to the naive MFE prediction of the focal sequence without alignment information. However, this approach is somewhat heavy handed and does not produce a perfect match with the ground truth.\n",
    "\n",
    "## Strategy II implementation (soft constraint)\n",
    "\n",
    "To refine the results further, we will implement **Strategy II**, which introduces a bonus energy as a soft constraint for each interaction. This bonus energy is proportional to the free energy difference in the ensemble with and without the interaction."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "90dc985d",
   "metadata": {},
   "source": [
    "First, calculate the base pair probability matrix of the ensemble. To do this, create a fold compound for the entire alignment and compute the probabilities from it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d076dd8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "9bcae19c",
   "metadata": {
    "type": "solution"
   },
   "outputs": [],
   "source": [
    "fc_aln = RNA.fold_compound(aligned_seq)\n",
    "css, cen = fc_aln.mfe()\n",
    "fc_aln.exp_params_rescale(cen)\n",
    "fc_aln.pf()\n",
    "\n",
    "bpp_matrix = np.array(fc_aln.bpp())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ceabfca",
   "metadata": {},
   "source": [
    "Similar to **Strategy I**, we now select all interactions that are compatible with the focal sequence and have a probability above a defined threshold. Unlike before, we can skip removing interactions which lead to hairpins of length < 3, as their energy penalty in the thermodynamic model is sufficient to outweigh any bonus energy we might assign.\n",
    "\n",
    "For this step, identify all interactions from the base pair probability matrix that:\n",
    "- Have a non-zero probability in the ensemble (you can define and experiment with a custom cutoff)\n",
    "- Do not involve gap characters in the focal sequence\n",
    "- Are formed by compatible nucleotides in the focal sequence\n",
    "\n",
    "To simplify the use of these results, store all valid interactions in an array. Each entry should include the indices of the two interacting partners and the corresponding base pair probability, e.g. `[[(4, 139), 0.997], ...]`\n",
    "\n",
    "*Hint: The base pair probability matrix is a square matrix, with non-zero elements only in its upper triangular part.\n",
    "You can also use the `plist` of the ViennaRNA-Package which  offers a very similar data structure. The function `plist_from_probs(cutoff)`, which is a member of `fold_compound`, can be called to create a plist which contains all basepairs with a probability above a given cutoff.*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed606f11",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "a2d0ef4a",
   "metadata": {
    "type": "solution"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[RNA.RNA.ep(i=4, j=139, p=0.99674, t=0), RNA.RNA.ep(i=8, j=135, p=0.999888, t=0), RNA.RNA.ep(i=18, j=75, p=0.979203, t=0), RNA.RNA.ep(i=20, j=71, p=0.99995, t=0), RNA.RNA.ep(i=25, j=68, p=0.999899, t=0), RNA.RNA.ep(i=33, j=62, p=0.373212, t=0), RNA.RNA.ep(i=36, j=57, p=0.987608, t=0), RNA.RNA.ep(i=84, j=122, p=0.935901, t=0), RNA.RNA.ep(i=85, j=121, p=0.921795, t=0), RNA.RNA.ep(i=94, j=114, p=0.996822, t=0), RNA.RNA.ep(i=95, j=113, p=0.995923, t=0), RNA.RNA.ep(i=96, j=112, p=0.997348, t=0), RNA.RNA.ep(i=97, j=111, p=0.999789, t=0), RNA.RNA.ep(i=101, j=108, p=0.987484, t=0), RNA.RNA.ep(i=102, j=107, p=0.988643, t=0)]\n"
     ]
    }
   ],
   "source": [
    "cutoff = 0.1\n",
    "interactions = [x for x in fc_aln.plist_from_probs(cutoff) if f\"{aligned_focal_seq[x.i]}{aligned_focal_seq[x.j]}\" in compatible_bases]\n",
    "print(interactions)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4441b940",
   "metadata": {},
   "source": [
    "In this step, we calculate the bonus energy for the soft constraints associated with each interaction.\n",
    "\n",
    "To compute the bonus energy for an interaction, calculate the ensemble free energy under two separate conditions:\n",
    "1. **Enforce the interaction**: Add a hard constraint to ensure the interaction is formed.\n",
    "2. **Forbid the interaction**: Add a hard constraint to prevent exactly this interaction from forming.\n",
    "\n",
    "The bonus energy for each interaction is then defined as the difference between the two ensemble free energies calculated above. Update each entry in the array of interactions by appending the corresponding bonus energy.\n",
    "\n",
    "*Hint: Use the `fc.hc_add_bp()` function to add hard constraints, along with the flags `RNA.CONSTRAINT_CONTEXT_ALL_LOOPS`, `RNA.CONSTRAINT_CONTEXT_ENFORCE`, and `RNA.CONSTRAINT_CONTEXT_NO_REMOVE`.*\n",
    "\n",
    "*You can either use two separate fold compounds for the two conditions or reset the constraints by calling `fc.hc_init()` after each calculation.*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d657a076",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "c5847f5d",
   "metadata": {
    "type": "solution"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[RNA.RNA.ep(i=4, j=139, p=-3.52698, t=0), RNA.RNA.ep(i=8, j=135, p=-5.60447, t=0), RNA.RNA.ep(i=18, j=75, p=-2.37402, t=0), RNA.RNA.ep(i=20, j=71, p=-6.0991, t=0), RNA.RNA.ep(i=25, j=68, p=-5.67049, t=0), RNA.RNA.ep(i=33, j=62, p=0.319538, t=0), RNA.RNA.ep(i=36, j=57, p=-2.69838, t=0), RNA.RNA.ep(i=84, j=122, p=-1.6524, t=0), RNA.RNA.ep(i=85, j=121, p=-1.52045, t=0), RNA.RNA.ep(i=94, j=114, p=-3.54275, t=0), RNA.RNA.ep(i=95, j=113, p=-3.38879, t=0), RNA.RNA.ep(i=96, j=112, p=-3.65455, t=0), RNA.RNA.ep(i=97, j=111, p=-5.21542, t=0), RNA.RNA.ep(i=101, j=108, p=-2.69219, t=0), RNA.RNA.ep(i=102, j=107, p=-2.75278, t=0)]\n"
     ]
    }
   ],
   "source": [
    "fc = RNA.fold_compound(aligned_seq)\n",
    "\n",
    "for interaction in interactions:\n",
    "\n",
    "    fc.hc_init()\n",
    "    fc.hc_add_bp(interaction.i, interaction.j, RNA.CONSTRAINT_CONTEXT_ALL_LOOPS | RNA.CONSTRAINT_CONTEXT_ENFORCE)\n",
    "    _, ener_w = fc.pf()\n",
    "\n",
    "    fc.hc_init()\n",
    "    fc.hc_add_bp(interaction.i, interaction.j, RNA.CONSTRAINT_CONTEXT_NO_REMOVE)\n",
    "    _, ener_wo = fc.pf()\n",
    "\n",
    "    # Overwrite the base pair probability by energy\n",
    "    interaction.p = ener_w-ener_wo\n",
    "\n",
    "print(interactions)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "69182ff8",
   "metadata": {},
   "source": [
    "Finally, create a fold compound using only the focal sequence and add the soft constraints based on the bonus energies calculated earlier. Assign the bonus energy to each interaction and determine the MFE structure. Use the `plot_structures(ss)` function to visualize the result compared to the ground truth structure.\n",
    "\n",
    "*Hint: Pay attention to the indices! The focal sequence used for folding does not include gap characters, whereas the base pair probability matrix uses indices from the alignment, which includes gap characters. Ensure proper mapping between the two.*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c30dd93",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "140449db",
   "metadata": {
    "type": "solution"
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr><td><img src='2_input/reference_plot.svg'></td><td><img src='2_input/result_plot9.svg'></td></tr>\n",
       "            <tr><td style='text-align: center;'>Reference</td><td style='text-align: center;'>Result</td></tr>\n",
       "        </table>\n",
       "    "
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fc = RNA.fold_compound(focal_seq)\n",
    "\n",
    "for interaction in interactions:\n",
    "    pos1 = interaction.i - aligned_focal_seq[:interaction.i].count('-')\n",
    "    pos2 = interaction.j - aligned_focal_seq[:interaction.j].count('-')\n",
    "\n",
    "    fc.sc_add_bp(pos1, pos2, interaction.p)\n",
    "\n",
    "ss, mfe = fc.mfe()\n",
    "\n",
    "plot_structures(ss)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
